### PR TITLE
Add PrivateCreator in table CustomTag

### PR DESCRIPTION
**Pull request by pengchen0692**
## Description
Add PrivateCreator in table CustomTag

## Related issues
Addresses AB380424

## Testing
Describe how this change was tested.

### CODE REVIEW

File: src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs [hunk 1]
```diff
@@ -94,14 +94,15 @@ public async Task<IReadOnlyList<CustomTagStoreEntry>> GetCustomTagsAsync(string
                 {
                     while (await reader.ReadAsync(cancellationToken))
                     {
-                        (int tagKey, string tagPath, string tagVR, int tagLevel, int tagStatus) = reader.ReadRow(
+                        (int tagKey, string tagPath, string tagVR, string tagPrivateCreator, int tagLevel, int tagStatus) = reader.ReadRow(
                             VLatest.CustomTag.TagKey,
                             VLatest.CustomTag.TagPath,
                             VLatest.CustomTag.TagVR,
+                            VLatest.CustomTag.TagPrivateCreator,
```
Comment (snehabandla) [inline src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs:101]: >VLatest.CustomTag.TagPrivateCreator [](start = 28, length = 35)

Won't this always be null?

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs [hunk 2]
```diff
@@ -35,6 +36,11 @@ public CustomTagStoreEntry(int key, string path, string vr, CustomTagLevel level
         /// </summary>
         public string VR { get; set; }
 
+        /// <summary>
+        /// Private Creator of this tag. Only apply to private tag.
```
Comment (snehabandla) [inline src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:40]: >/// Private Creator of this tag. Only apply to private tag. [](start = 8, length = 59)

Can you explain what a private creator is and what the use of this field is?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 3]
```diff
@@ -369,12 +369,14 @@ CREATE NONCLUSTERED INDEX IX_ChangeFeed_StudyInstanceUid_SeriesInstanceUid_SopIn
     Stores added custom tags
     TagPath is represented without any delimiters and each level takes 8 bytes
     TagLevel can be 0, 1 or 2 to represent Instance, Series or Study level
+    TagPrivateCreator is private creator, only apply to private tag.
     TagStatus can be 0, 1 or 2 to represent Reindexing, Added or Deindexing
```
Comment (snehabandla) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:373]: Here too, please add  description on what this value actually represents.

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs (hunk 2 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:40]: Hi Sneha, this is the "Private Creator" I'm talking about in today's syncup.
Private Tag is widely used in industry, it likely to happen that 2 users are using dicom tag of same groupId and elementId, but means different. The private creator is description of implementer, could help distinguish that. 
More details can be find in standard:  http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.8.html

Comment (snehabandla) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:40]: Sorry if that was unclear - i meant add a description in the summary!

---
In reply to: [595574340](https://github.com/microsoft/dicom-server/pull/644#discussion_r595574340) [](ancestors = 595574340)

Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:40]: I will add these information to summary.

File: src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs (hunk 1 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs:101]: it is Null for standard tags, should be some value for private tags.
I'll update AddCustomTag api to require privateCreator for private tag.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql (hunk 3 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:373]: will add a brief description here.

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs [hunk 4]
```diff
@@ -10,11 +10,12 @@ namespace Microsoft.Health.Dicom.Core.Features.CustomTag
     /// </summary>
     public class CustomTagStoreEntry
     {
-        public CustomTagStoreEntry(int key, string path, string vr, CustomTagLevel level, CustomTagStatus status)
+        public CustomTagStoreEntry(int key, string path, string vr, string privateCreator, CustomTagLevel level, CustomTagStatus status)
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:13]: All of the `CustomTagStoreEntries` aren't private are they? So should this optional?

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs [hunk 5]
```diff
@@ -35,6 +36,13 @@ public CustomTagStoreEntry(int key, string path, string vr, CustomTagLevel level
         /// </summary>
         public string VR { get; set; }
 
+        /// <summary>
+        /// Private Creator of this tag.
+        /// It's identification code of private tag implementer, only apply to private tag.
+        /// Please refer to http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.8.html for more details.
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:42]: The comment starting at `It's identification code...` should be in a `<remarks>...</remarks>` element

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs [hunk 6]
```diff
@@ -35,6 +36,13 @@ public CustomTagStoreEntry(int key, string path, string vr, CustomTagLevel level
         /// </summary>
         public string VR { get; set; }
 
+        /// <summary>
+        /// Private Creator of this tag.
+        /// It's identification code of private tag implementer, only apply to private tag.
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:41]: It's the*

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs (hunk 4 above)
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:13]: Should there be more validation of the input?

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/GetCustomTagsServiceTests.cs [hunk 7]
```diff
@@ -42,8 +42,8 @@ public async Task GivenRequestForAllTags_WhenNoTagsAreStored_ThenExceptionShould
         [Fact]
         public async Task GivenRequestForAllTags_WhenMultipleTagsAreStored_ThenCustomTagEntryListShouldBeReturned()
         {
-            CustomTagStoreEntry tag1 = CreateCustomTagEntry(1, "45456767", DicomVRCode.AE.ToString(), CustomTagLevel.Instance, CustomTagStatus.Added);
-            CustomTagStoreEntry tag2 = CreateCustomTagEntry(2, "01012323", DicomVRCode.FL.ToString(), CustomTagLevel.Series, CustomTagStatus.Reindexing);
+            CustomTagStoreEntry tag1 = CreateCustomTagEntry(1, "45456767", DicomVRCode.AE.ToString(), null, CustomTagLevel.Instance, CustomTagStatus.Added);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/GetCustomTagsServiceTests.cs:45]: We should probably have an example where the private tag creator is supplied

**Comment by wsugarman** [comment]
Is the purpose of this change to provide a unique key if the group number is reused?

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs (hunk 4 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:13]: I think it's fine to put it in the end. my perspective is CustomTagStoreEntry represent an entry in database, so just mapping to whatever in database.

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs (hunk 6 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:41]: will fix it.

File: src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs (hunk 4 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagStoreEntry.cs:13]: CustomTagStoreEntry just stand for a single row in database.

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/GetCustomTagsServiceTests.cs (hunk 7 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/GetCustomTagsServiceTests.cs:45]: * This task focus on adding new column in database, coming task will add more unit test for coverage
* This unit test verifies GetCustomTagService API returns correct data, and without coming task (https://microsofthealth.visualstudio.com/Health/_boards/board/t/Medical%20Imaging/Stories/?workitem=80425), we wont' be able to do it.

**Comment by pengchen0692** [comment]
> Is the purpose of this change to provide a unique key if the group number is reused?

this can mitigate,  but cannot perfectly solve the problem.

Private Tag is widely used in industry, it likely to happen that multiple users are using dicom tag of same groupId and elementId, but means different. The private creator is description of implementer, could help distinguish that.
More details can be find in standard: http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.8.html

**Review by wsugarman — APPROVED** [review]

**Pull request merged**